src/main/java/com/epam/aidial/core/storage/BlobWriteStream.java

```diff
@@ -0,0 +1,187 @@
+                if (exception != null) {
+                    throw new RuntimeException(exception);
+                }
+
+                Buffer lastChunk = chunkBuffer.slice(0, position);
+                metadata = new FileMetadata(fileName, BlobStorageUtil.removeTrailingPathSeparator(parentPath),
+                        bytesHandled, contentType);
+                if (mpu == null) {
+                    log.info("Resource is too small for multipart upload, sending as a regular blob");
+                    storage.store(fileName, parentPath, contentType, lastChunk);
+                } else {
+                    if (position != 0) {
+                        MultipartPart part = storage.storeMultipartPart(mpu, ++chunkNumber, lastChunk);
+                        parts.add(part);
+                    }
+
+                    storage.completeMultipartUpload(mpu, parts);
+                    log.info("Multipart upload committed, bytes handled {}", bytesHandled);
+                }
+
+                return null;
+            }
+        });
+        result.onSuccess(success -> {
+            if (handler != null) {
+                handler.handle(Future.succeededFuture());
+            }
+        }).onFailure(error -> {
+            if (handler != null) {
+                handler.handle(Future.failedFuture(error));
+            }
+        });
+    }
+
+    @Override
+    public synchronized WriteStream<Buffer> setWriteQueueMaxSize(int maxSize) {
+        assert maxSize > MIN_PART_SIZE;
+        chunkSize = maxSize;
+        return this;
+    }
+
+    @Override
+    public synchronized boolean writeQueueFull() {
+        return isBufferFull;
+    }
+
+    @Override
+    public WriteStream<Buffer> drainHandler(Handler<Void> handler) {
+        vertx.executeBlocking(() -> {
+            synchronized (BlobWriteStream.this) {
+                try {
+                    if (mpu == null) {
+                        mpu = storage.initMultipartUpload(fileName, parentPath, contentType);
+                    }
+                    MultipartPart part = storage.storeMultipartPart(mpu, ++chunkNumber, chunkBuffer.slice(0, position));
+                    parts.add(part);
+                    position = 0;
+                    isBufferFull = false;
+                } catch (Throwable ex) {
+                    exception = ex;
+                } finally {
+                    if (handler != null) {
+                        handler.handle(null);
+                    }
+                }
+            }
+            return null;
+        });
+
+        return this;
+    }
+
+    public synchronized FileMetadata getMetadata() {
+        return metadata;
+    }
+
+    public synchronized void abortUpload(Throwable ex) {
+        if (mpu != null) {
+            storage.abortMultipartUpload(mpu);
+        }
+
+        if (errorHandler != null) {
+            errorHandler.handle(ex);
+        }
+
+        log.info("Multipart upload aborted");
```

Comment:
the log level should be at least warning

Reply:
fixed